src/cljam/io/sam/util/validator.clj

```diff
@@ -0,0 +1,191 @@
+           res1
+           (list* res2 res3 more))))
+
+(defn- validate-pos* [rname pos refmap]
+  (let [max-len (get-in refmap [rname :LN])]
+    (cond
+      (not (integer? pos)) ["Must be integer."]
+      (not (<= 0 (int pos) Integer/MAX_VALUE))
+      ["Must be in the [0, 2147483647]."]
+
+      (and max-len
+           (> (int pos) (int max-len)))
+      [(format "Must be less than or equal %d." max-len)])))
+
+(defn- validate-pos [{:keys [refmap]} {:keys [pos rname]}]
+  (when-let [err (validate-pos* rname pos refmap)]
+    (apply error :pos err)))
+
+(defn- validate-pnext [{:keys [refmap]} {:keys [pnext rname]}]
+  (when-let [err (validate-pos* rname pnext refmap)]
+    (apply error :pnext err)))
+
+(defn- validate-rname* [rname refmap]
+  (cond
+    (not (string? rname)) ["Must be string."]
+    (and (not (= rname "*"))
+         (not (get refmap rname)))
+    [(format "Must be not in header.(%s)" rname)]))
+
+(defn- validate-rname [{:keys [refmap]} {:keys [rname]}]
+  (when-let [err (validate-rname* rname refmap)]
+    (apply error :rname err)))
+
+(defn- validate-rnext [{:keys [refmap]} {:keys [rnext]}]
+  (when-let [err (validate-rname* rnext refmap)]
+    (error :rnext err)))
+
+(defn- validate-qname [_ {:keys [qname]}]
+  (if (not (string? qname))
+    (error :qname "Must be string.")
+    (when-let [res
+               (cond-> nil
+                 (not (<= (count qname) 254))
+                 (conj "Must be less than or equal to 254 characters.")
+
+                 (not (re-matches #"^[!-?A-~]+$" qname))
+                 (conj "Must not contain illegal characters."))]
+      {:errors {:qname res}})))
+
+(defn- validate-mapq [_ {:keys [mapq]}]
+  (cond
+    (not (integer? mapq)) (error :mapq "Must be integer.")
+    (not (<= 0 (int mapq) 255))
+    (error :mapq "Must be in the [0-255].")))
+
+(defn- validate-cigar [_ {:keys [cigar]}]
+  (cond
+    (not (string? cigar)) (error :cigar "Must be string.")
+    (not (re-matches #"^\*|([0-9]+[MIDNSHPX=])+$" cigar))
+    (error :cigar "Invalid format.")))
+
+(defn- validate-tlen [_ {:keys [tlen]}]
+  (cond
+    (not (integer? tlen)) (error :tlen "Must be integer.")
+    (not (<= (- Integer/MAX_VALUE) tlen Integer/MAX_VALUE))
+    (error :tlen "Must be in the [-2147483647,2147483647].")))
+
+(defn- validate-qual [_ {:keys [qual]}]
+  (cond
+    (not (string? qual)) (error :qual "Must be string.")
+    (not (re-matches #"[!-~]+" qual)) (error :qual "Must not contain bad character.")))
+
+(defn- validate-seq [_ {:keys [seq]}]
+  (cond
+    (not (string? seq)) (error :seq "Must be string.")
+    (not (re-matches #"\*|[A-Za-z=.]+" seq)) (error :seq "Must not contain bad character.")))
+
+(defn- validate-option [{:keys [type value]}]
+  (case type
+    "A" (when-not (and (char? value) (<= (int \!) (int value) (int \~)))
+          ["Must be char [!-~]."])
+    "i" (when-not (and (integer? value) (<= -32767 value 32767))
+          ["Must be 16 bit signed integer."])
```

Comment:
In SAM, the only TYPE `i` can be used for an integer (practically in the range $`[-2^{31},2^{32})`$)
In BAM, any of `cCsSiI` may be used.